pkg/segment/aggregations/segaggs.go

```diff
@@ -196,6 +209,109 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
 	return nil
 }
 
+func performTail(nodeResult *structs.NodeResult, tailExpr *structs.TailExpr, recs map[string]map[string]interface{}, recordIndexInFinal map[string]int, finishesSegment bool, numTotalSegments uint64, hasSort bool) error {
+
+	if nodeResult.Histogram != nil {
+		for _, aggResult := range nodeResult.Histogram {
+			diff := len(aggResult.Results) - int(tailExpr.TailRows)
+			if diff > 0 {
+				aggResult.Results = aggResult.Results[diff:]
+				tailExpr.TailRows = 0
+			} else {
+				tailExpr.TailRows -= uint64(len(aggResult.Results))
+			}
+			n := len(aggResult.Results)
+			for i := 0; i < n/2; i++ {
+				aggResult.Results[i], aggResult.Results[n-i-1] = aggResult.Results[n-i-1], aggResult.Results[i]
+			}
+			if tailExpr.TailRows == 0 {
+				break
+			}
+		}
+
+		return nil
+	}
+
+	if finishesSegment {
+		tailExpr.NumProcessedSegments++
+	}
+
+	if tailExpr.TailRecords == nil {
+		tailExpr.TailRecords = make(map[string]map[string]interface{}, 0)
+	}
+	if tailExpr.TailPQ == nil {
+		pq := make(utils.PriorityQueue, 0)
+		tailExpr.TailPQ = &pq
+		heap.Init(tailExpr.TailPQ)
+	}
+
+	if !hasSort {
+		for k, record := range recs {
+			timeVal, exists := record["timestamp"]
+			if !exists {
+				continue
+			}
+			heap.Push(tailExpr.TailPQ, &utils.Item{
+				Priority: float64(timeVal.(uint64)),
+				Value:    k,
+			})
+			tailExpr.TailRecords[k] = record
+			if tailExpr.TailPQ.Len() > int(tailExpr.TailRows) {
+				item := heap.Pop(tailExpr.TailPQ).(*utils.Item)
+				delete(tailExpr.TailRecords, item.Value)
+			}
+			delete(recs, k)
+		}
+	}
+
+	if tailExpr.NumProcessedSegments < numTotalSegments {
+		if hasSort && len(recs) > 0 {
+			log.Errorf("performTail: Sort was applied but still records are found in recs for a non-last segment")
+		}
+		return nil
```

Comment:
You should probably also return an error when you log the error

Reply:
addressed in the new commit. Just returning the error because we are logging that same error in postQueryBucketCleaning. So, no point in logging it twice.